packages/lib-classifier/src/store/FeedbackStore.js

```diff
@@ -1,45 +1,76 @@
 import { autorun } from 'mobx'
-import { addDisposer, getRoot, onAction, types } from 'mobx-state-tree'
+import { addDisposer, addMiddleware, getRoot, onAction, types } from 'mobx-state-tree'
+import { flatten } from 'lodash'
 
 import helpers from './feedback/helpers'
 import strategies from './feedback/strategies'
 
 const FeedbackStore = types
   .model('FeedbackStore', {
     isActive: types.optional(types.boolean, false),
-    rules: types.map(types.frozen({}))
+    rules: types.map(types.frozen({})),
+    showModal: types.optional(types.boolean, false),
+    allowSubjectAdvance: types.optional(types.boolean, false)
   })
+  .views(self => ({
+    get hideSubjectViewer () {
+      return flatten(Array.from(self.rules.values()))
+        .some(rule => rule.hideSubjectViewer)
+    },
+    get messages () {
+      return flatten(Array.from(self.rules.values()))
+        .map(rule => {
+          if (rule.success && rule.successEnabled) {
+            return rule.successMessage
+          } else if (!rule.success && rule.failureEnabled) {
+            return rule.failureMessage
+          }
+        })
+    }
+  }))
   .actions(self => {
     function afterAttach () {
-      createSubjectObserver()
       createClassificationObserver()
-    }
-
-    function createSubjectObserver () {
-      const subjectDisposer = autorun(() => {
-        const subject = getRoot(self).subjects.active
-        if (subject) {
-          self.reset()
-          self.createRules(subject)
-        }
-      })
-      addDisposer(self, subjectDisposer)
+      createSubjectMiddleware()
+      createSubjectObserver()
     }
 
     function createClassificationObserver () {
       const classificationDisposer = autorun(() => {
         onAction(getRoot(self).classifications, (call) => {
           if (call.name === 'completeClassification') {
             const annotations = getRoot(self).classifications.currentAnnotations
-            for (const value of annotations.values()) {
-              self.update(value)
-            }
+            annotations.forEach(annotation => self.update(annotation))
           }
         })
       })
       addDisposer(self, classificationDisposer)
     }
 
+    function createSubjectMiddleware () {
+      const subjectMiddleware = autorun(() => {
+        addMiddleware(getRoot(self).subjects, (call, next, abort) => {
+          if (call.name === 'advance' && self.isActive && self.messages.length && !self.allowSubjectAdvance) {
```

Comment:
Reading this again, I'd move the logic that's specific to feedback (`self.isActive && self.messages.length`) into the `showFeedback` action. It feels awkward to have that check here.

Reply:
Also, the generic name `subjectMiddleware` should be changed to describe what this function does. Would `onSubjectsAdvance` describe what the function does?